File: src/main/java/com/hp/jipp/encoding/CollectionType.java
package com.hp.jipp.encoding;

import com.google.common.collect.ImmutableList;
import com.hp.jipp.util.ParseError;
import com.hp.jipp.util.Util;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/**
 * A type for attribute collections.
 *
 * @see <a href="https://tools.ietf.org/html/rfc3382">RFC3382</a>
 */
public class CollectionType extends AttributeType<AttributeCollection> {
    private static final String TYPE_NAME = "Collection";

    /** Used to terminate a collection */
    private static final Attribute<byte[]> EndCollectionAttribute = new OctetStringType(Tag.EndCollection, "").of();

    static final Attribute.BaseEncoder<AttributeCollection>
            ENCODER = new Attribute.BaseEncoder<AttributeCollection>() {

        @Override
        public String getType() {
            return TYPE_NAME;
        }

        @Override
        public void writeValue(DataOutputStream out, AttributeCollection value)
                throws IOException {
            out.writeShort(0); // Empty value

            for (Attribute<?> attribute : value.getAttributes()) {
                // Write a MemberAttributeName attribute
                Tag.MemberAttributeName.write(out);
                out.writeShort(0);
                Attribute.writeValueBytes(out, attribute.getName().getBytes(Util.UTF8));

                // Write the attribute, but without its name
                attribute.withName("").write(out);
            }

            // Terminating attribute
            EndCollectionAttribute.write(out);
        }

        @Override
        public AttributeCollection readValue(DataInputStream in, Attribute.EncoderFinder finder, Tag valueTag)
                throws IOException {
            Attribute.skipValueBytes(in);
            ImmutableList.Builder<Attribute<?>> builder = new ImmutableList.Builder<>();

            // Read attribute pairs until EndCollection is reached.
            while (true) {
                Tag tag = Tag.read(in);
                if (tag == Tag.EndCollection) {
                    // Skip the rest of this attr and return.
                    Attribute.skipValueBytes(in);
                    Attribute.skipValueBytes(in);
                    break;
                } else if (tag == Tag.MemberAttributeName) {
                    Attribute.skipValueBytes(in);
                    String memberName = new String(Attribute.Companion.readValueBytes(in), Util.UTF8);
                    Tag memberTag = Tag.read(in);

                    // Read and throw away the blank attribute name
                    Attribute.readValueBytes(in);
                    builder.add(finder.find(memberTag, memberName).read(in, finder, memberTag, memberName));

                } else {
                    throw new ParseError("Bad tag in collection: " + tag);
                }
            }
            return new AttributeCollection(builder.build());
        }

        @Override
        public boolean valid(Tag valueTag) {
            return valueTag == Tag.BeginCollection;
        }
    };

    public CollectionType(String name) {
        super(ENCODER, Tag.BeginCollection, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/CollectionType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import com.hp.jipp.util.ParseError

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/**
 * A type for attribute collections.

 * See RFC3382 (https://tools.ietf.org/html/rfc3382)
 */
open class CollectionType(override val name: String) :
        AttributeType<AttributeCollection>(Encoder, Tag.beginCollection) {

    /** Return a collection containing the supplied attributes */
    fun of(vararg attributes: Attribute<*>) = this(AttributeCollection(attributes.toList()))

    companion object Encoder : com.hp.jipp.encoding.Encoder<AttributeCollection>() {
        override val typeName
            get() = "Collection"

        // Terminates a collection
        private val endCollectionAttribute = OctetStringType(Tag.endCollection, "").empty()

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: AttributeCollection) {
            out.writeShort(0) // Empty value

            for (attribute in value.attributes) {
                // Write a memberAttributeName attribute
                Tag.memberAttributeName.write(out)
                out.writeShort(0)
                out.writeString(attribute.name)

                // Write the attribute, but without its name
                attribute.withName("").write(out)
            }

            // Terminating attribute
            endCollectionAttribute.write(out)
        }

        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, finder: Finder, valueTag: Tag): AttributeCollection {
            input.skipValueBytes()
            val builder = ArrayList<Attribute<*>>()

            // Read attribute pairs until endCollection is reached.
            while (true) {
                val tag = Tag.read(input)
                if (tag === Tag.endCollection) {
                    // Skip the rest of this attr and return.
                    input.skipValueBytes()
                    input.skipValueBytes()
                    break
                } else if (tag === Tag.memberAttributeName) {
                    input.skipValueBytes()
                    val memberName = input.readString()
                    val memberTag = Tag.read(input)

                    // Read and throw away the blank attribute name
                    input.readValueBytes()
                    val encoder = finder.find(memberTag, memberName)
                    builder.add(input.readAttribute(encoder, finder, memberTag, memberName))

                } else {
                    throw ParseError("Bad tag in collection: $tag")
                }
            }
            return AttributeCollection(builder)
        }

        override fun valid(valueTag: Tag): Boolean {
            return valueTag === Tag.beginCollection
        }
    }
}
